sdk/rust/oak_runtime_compat/src/lib.rs

```diff
@@ -0,0 +1,68 @@
+//
+// Copyright 2019 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! Functionality to allow use of Rust code within the Oak Runtime.
+
+#![no_std]
+
+#![feature(lang_items)]
+#![feature(allocator_api)]
+#![feature(alloc_prelude)]
+#![feature(alloc_error_handler)]
+#![feature(fn_traits)]
+#![feature(rustc_private)]
+#![feature(never_type)]
+#![feature(box_syntax)]
+#![feature(int_error_internals)]
+#![feature(array_error_internals)]
+#![feature(char_error_internals)]
+
+extern crate alloc;
+extern crate core;
+extern crate libc;
+
+use alloc::prelude::v1::*;
+
+pub mod asylo;
+pub mod common;
+
+#[global_allocator]
+static A: asylo::allocator::System = asylo::allocator::System;
+
+pub fn thread_test() -> common::io::Result<i32> {
+  use alloc::sync::Arc;
+  use core::sync::atomic::{AtomicI32, Ordering};
+  let val = Arc::new(AtomicI32::new(2));
+
+  let other = {
+    let val = Arc::clone(&val);
+    move || {
+      val.fetch_add(40, Ordering::SeqCst);
+    }
+  };
+
+  let t = unsafe {
+      asylo::thread::Thread::new(Box::new(other))
```

Comment:
I guess this API should also have a safe wrapper at some point, could you add a TODO for it?

Reply:
Yup good idea

Reply:
Probably the TODO should also / instead be on `Thread::new`? -- actually, it seems that the "safe" API for threads in Rust std is actually this https://github.com/rust-lang/rust/blob/6b561b4917e803c4be4ca44d8e552b680cb9e380/src/libstd/thread/mod.rs#L448 ? Though I haven't actually looked at why that is.

Reply:
Basically they add thread naming and argument passing. The thread naming won't work through the asylo posix interface, the argument passing with safe interface I just left temporarily until some of this was merged